osmosis-labs/osmosis
# fix: adding lazy migration for cl incentives

Closes: #XXX

## What is the purpose of the change

This PR migrates the rest of the cl pools to the new scaling factor, this PR makes sure that we follow the same logic for the CL incentive logic for all pools

### How to test

To test this PR you need to use the `in-place-testnet` 

Download the latest snapshot and genesis

```bash
OSMOSIS_HOME=$HOME/.osmosisd

SNAPSHOT_URL=$(curl -sL https://snapshots.osmosis.zone/latest)
RPC_URL=https://rpc.osmosis.zone
ADDRBOOK_URL="https://rpc.osmosis.zone/addrbook"
GENESIS_URL=https://github.com/osmosis-labs/osmosis/raw/main/networks/osmosis-1/genesis.json

rm -rf $HOME/.osmosisd/

osmosisd init test --chain-id testing --home=$HOME/.osmosisd

# Copy genesis
echo -e "\nDownloading genesis file..."
wget $GENESIS_URL -O $OSMOSIS_HOME/config/genesis.json
echo ✅ Genesis file downloaded successfully.

# Download latest snapshot
echo -e "\nDownloading latest snapshot..."
wget -O - $SNAPSHOT_URL | lz4 -d | tar -C $OSMOSIS_HOME/ -xf -
echo -e ✅ Snapshot downloaded successfully.

# Run the node
osmosisd start --home=$HOME/.osmosisd
```

You need to install v23.0.8 to run the following command
```
osmosisd in-place-testnet edgenet osmo12smx2wdlyttvyzvzg54y2vnqwq2qjateuf7thj --trigger-testnet-upgrade v24
```

Wait until the node stops, build this branch and run
```
osmosisd start
```

This will run the v24 migration and migrate all remaining CL pool incentive accumlators

In the below are all the accumlators before and after the migration:

https://docs.google.com/spreadsheets/d/1CnhEKEjwOBKpsNYWrbwhpyfuS9Yva_qhhoCmJ6sbjJc/edit#gid=1264576604

How to test manually:

- Create an incentive for a pool
- Create a CL position in a tight range
- Swap tokens
- Check that rewards have been distributed

How to run unit test:
- `cd ./x/concentrated-liquidity`
- `go test ./... -run TestKeeperTestSuite/TestBeginBlockLazyMigration`


TODO:
- [x] Finish writing tests for lazy migration
